In [148]:
import pandas as pd

In [149]:
# Loading the dataset
df = pd.read_csv("lendingclub_raw.csv", low_memory=False)

In [150]:
df.head()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,NaN,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,68355089,NaN,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2,68341763,NaN,20000.0,20000.0,20000.0,60 months,10.78,432.66,B,B4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
3,66310712,NaN,35000.0,35000.0,35000.0,60 months,14.85,829.90,C,C5,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
4,68476807,NaN,10400.0,10400.0,10400.0,60 months,22.45,289.91,F,F1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN


In [151]:
df.shape

(658974, 151)

In [152]:
df.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       ...
       'hardship_payoff_balance_amount', 'hardship_last_payment_amount',
       'disbursement_method', 'debt_settlement_flag',
       'debt_settlement_flag_date', 'settlement_status', 'settlement_date',
       'settlement_amount', 'settlement_percentage', 'settlement_term'],
      dtype='object', length=151)

In [153]:
# To clearly see all column names - storing in a list.
lst = df.columns.tolist()
lst

['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'annual_inc_joint',
 '

In [154]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 658974 entries, 0 to 658973
Columns: 151 entries, id to settlement_term
dtypes: float64(113), object(38)
memory usage: 759.2+ MB


In [155]:
# For business analysis, borrower profile and loan details are needed.
columns_needed = ["loan_amnt",
                  "term",
                  "int_rate",
                  "grade",
                  "emp_length",
                  "home_ownership",
                  "annual_inc",
                  "loan_status",
                  "purpose",
                  "dti",
                  "fico_range_low",
                  "fico_range_high"]

In [156]:
# Analysis to be done in a dataframe with selective columns & their data
df = df[columns_needed].copy()

In [157]:
# After selecting relevant loan and borrower attributes, Exploratory checks are to be performed using info(), describe(),
# and null analysis to understand data quality, distributions, and required cleaning steps.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 658974 entries, 0 to 658973
Data columns (total 12 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_amnt        658968 non-null  float64
 1   term             658968 non-null  object 
 2   int_rate         658968 non-null  float64
 3   grade            658968 non-null  object 
 4   emp_length       616717 non-null  object 
 5   home_ownership   658968 non-null  object 
 6   annual_inc       658968 non-null  float64
 7   loan_status      658967 non-null  object 
 8   purpose          658967 non-null  object 
 9   dti              658519 non-null  float64
 10  fico_range_low   658967 non-null  float64
 11  fico_range_high  658967 non-null  float64
dtypes: float64(6), object(6)
memory usage: 60.3+ MB


In [158]:
df.describe()

,loan_amnt,int_rate,annual_inc,dti,fico_range_low,fico_range_high
count,658968.000000,658968.000000,6.589680e+05,658519.000000,658967.000000,658967.000000
mean,15250.678068,12.746364,7.760937e+04,19.307485,697.144887,701.145082
std,9066.521225,4.663640,1.542633e+05,14.406556,32.600108,32.600992
min,1000.000000,5.310000,0.000000e+00,0.000000,660.000000,664.000000
25%,8000.000000,9.430000,4.600000e+04,12.280000,670.000000,674.000000
50%,13500.000000,12.290000,6.500000e+04,18.370000,690.000000,694.000000
75%,20000.000000,15.590000,9.220853e+04,25.210000,715.000000,719.000000
max,40000.000000,30.990000,1.100000e+08,999.000000,845.000000,850.000000


In [159]:
df.isnull().sum()

,0
loan_amnt,6
term,6
int_rate,6
grade,6
emp_length,42257
home_ownership,6
annual_inc,6
loan_status,7
purpose,7
dti,455


In [160]:
df.head()

,loan_amnt,term,int_rate,grade,emp_length,home_ownership,annual_inc,loan_status,purpose,dti,fico_range_low,fico_range_high
0,3600.0,36 months,13.99,C,10+ years,MORTGAGE,55000.0,Fully Paid,debt_consolidation,5.91,675.0,679.0
1,24700.0,36 months,11.99,C,10+ years,MORTGAGE,65000.0,Fully Paid,small_business,16.06,715.0,719.0
2,20000.0,60 months,10.78,B,10+ years,MORTGAGE,63000.0,Fully Paid,home_improvement,10.78,695.0,699.0
3,35000.0,60 months,14.85,C,10+ years,MORTGAGE,110000.0,Current,debt_consolidation,17.06,785.0,789.0
4,10400.0,60 months,22.45,F,3 years,MORTGAGE,104433.0,Fully Paid,major_purchase,25.37,695.0,699.0


In [161]:
# Cleaning Loan Amount
df["loan_amnt"].isnull().sum()

np.int64(6)

In [162]:
# There are 10 null values & since it's a critical data we require for analysyis - dropping the null rows
df=df.dropna(subset=["loan_amnt"])

In [163]:
df.isnull().sum()
# Many null value rows got removed with it.

,0
loan_amnt,0
term,0
int_rate,0
grade,0
emp_length,42251
home_ownership,0
annual_inc,0
loan_status,1
purpose,1
dti,449


In [164]:
#Term
df.term.unique()

array([' 36 months', ' 60 months'], dtype=object)

In [165]:
# Remove uneccessary space
df["term"] = df["term"].str.strip()

In [166]:
# Remove months string from data
df["term"] = df["term"].str.replace(" months","").astype(int)

In [167]:
# Rename the term attribute
df = df.rename(columns={"term":"term_in_months"})

In [168]:
# Employee length
df["emp_length"] = df["emp_length"].fillna("Unknown")

In [169]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 658968 entries, 0 to 658973
Data columns (total 12 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_amnt        658968 non-null  float64
 1   term_in_months   658968 non-null  int64  
 2   int_rate         658968 non-null  float64
 3   grade            658968 non-null  object 
 4   emp_length       658968 non-null  object 
 5   home_ownership   658968 non-null  object 
 6   annual_inc       658968 non-null  float64
 7   loan_status      658967 non-null  object 
 8   purpose          658967 non-null  object 
 9   dti              658519 non-null  float64
 10  fico_range_low   658967 non-null  float64
 11  fico_range_high  658967 non-null  float64
dtypes: float64(6), int64(1), object(5)
memory usage: 65.4+ MB


In [170]:
# Loan Status
df["loan_status"].value_counts()

,count
loan_status,
Fully Paid,354806
Current,202949
Charged Off,91907
Late (31-120 days),5992
In Grace Period,2212
Late (16-30 days),1090
Default,11


In [171]:
# When applied for loan - system usually decides wheather accept it or reject not too many things to decide based on many criterias
df ["decision"] = df["loan_status"].apply(lambda x: "Approved" if x in ["Fully Paid", "Current"] else "Rejected")

In [172]:
df["decision"].value_counts()

,count
decision,
Approved,557755
Rejected,101213


In [173]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 658968 entries, 0 to 658973
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_amnt        658968 non-null  float64
 1   term_in_months   658968 non-null  int64  
 2   int_rate         658968 non-null  float64
 3   grade            658968 non-null  object 
 4   emp_length       658968 non-null  object 
 5   home_ownership   658968 non-null  object 
 6   annual_inc       658968 non-null  float64
 7   loan_status      658967 non-null  object 
 8   purpose          658967 non-null  object 
 9   dti              658519 non-null  float64
 10  fico_range_low   658967 non-null  float64
 11  fico_range_high  658967 non-null  float64
 12  decision         658968 non-null  object 
dtypes: float64(6), int64(1), object(6)
memory usage: 70.4+ MB


In [174]:
# DTI
df["dti"].isnull().sum()

np.int64(449)

In [175]:
df["dti"]=df["dti"].fillna(df["dti"].median())

In [176]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 658968 entries, 0 to 658973
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_amnt        658968 non-null  float64
 1   term_in_months   658968 non-null  int64  
 2   int_rate         658968 non-null  float64
 3   grade            658968 non-null  object 
 4   emp_length       658968 non-null  object 
 5   home_ownership   658968 non-null  object 
 6   annual_inc       658968 non-null  float64
 7   loan_status      658967 non-null  object 
 8   purpose          658967 non-null  object 
 9   dti              658968 non-null  float64
 10  fico_range_low   658967 non-null  float64
 11  fico_range_high  658967 non-null  float64
 12  decision         658968 non-null  object 
dtypes: float64(6), int64(1), object(6)
memory usage: 70.4+ MB


In [177]:
# Creating Credit Score
df["credit_score"] = (df["fico_range_low"] + df["fico_range_high"]) / 2

In [178]:
# Creating Risk Segment
df["risk_segment"] = df.apply(lambda x: "Low Risk" if (x["credit_score"] >= 700) or (x["dti"] < 25) else "High Risk", axis = 1)

In [179]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 658968 entries, 0 to 658973
Data columns (total 15 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_amnt        658968 non-null  float64
 1   term_in_months   658968 non-null  int64  
 2   int_rate         658968 non-null  float64
 3   grade            658968 non-null  object 
 4   emp_length       658968 non-null  object 
 5   home_ownership   658968 non-null  object 
 6   annual_inc       658968 non-null  float64
 7   loan_status      658967 non-null  object 
 8   purpose          658967 non-null  object 
 9   dti              658968 non-null  float64
 10  fico_range_low   658967 non-null  float64
 11  fico_range_high  658967 non-null  float64
 12  decision         658968 non-null  object 
 13  credit_score     658967 non-null  float64
 14  risk_segment     658968 non-null  object 
dtypes: float64(7), int64(1), object(7)
memory usage: 80.4+ MB


In [180]:
df.head()

,loan_amnt,term_in_months,int_rate,grade,emp_length,home_ownership,annual_inc,loan_status,purpose,dti,fico_range_low,fico_range_high,decision,credit_score,risk_segment
0,3600.0,36,13.99,C,10+ years,MORTGAGE,55000.0,Fully Paid,debt_consolidation,5.91,675.0,679.0,Approved,677.0,Low Risk
1,24700.0,36,11.99,C,10+ years,MORTGAGE,65000.0,Fully Paid,small_business,16.06,715.0,719.0,Approved,717.0,Low Risk
2,20000.0,60,10.78,B,10+ years,MORTGAGE,63000.0,Fully Paid,home_improvement,10.78,695.0,699.0,Approved,697.0,Low Risk
3,35000.0,60,14.85,C,10+ years,MORTGAGE,110000.0,Current,debt_consolidation,17.06,785.0,789.0,Approved,787.0,Low Risk
4,10400.0,60,22.45,F,3 years,MORTGAGE,104433.0,Fully Paid,major_purchase,25.37,695.0,699.0,Approved,697.0,High Risk


In [181]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 658968 entries, 0 to 658973
Data columns (total 15 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_amnt        658968 non-null  float64
 1   term_in_months   658968 non-null  int64  
 2   int_rate         658968 non-null  float64
 3   grade            658968 non-null  object 
 4   emp_length       658968 non-null  object 
 5   home_ownership   658968 non-null  object 
 6   annual_inc       658968 non-null  float64
 7   loan_status      658967 non-null  object 
 8   purpose          658967 non-null  object 
 9   dti              658968 non-null  float64
 10  fico_range_low   658967 non-null  float64
 11  fico_range_high  658967 non-null  float64
 12  decision         658968 non-null  object 
 13  credit_score     658967 non-null  float64
 14  risk_segment     658968 non-null  object 
dtypes: float64(7), int64(1), object(7)
memory usage: 80.4+ MB


# Analytics

1. Overall Loan Approval Rate

In [182]:
df["decision"].value_counts(normalize=True)*100

,proportion
decision,
Approved,84.640681
Rejected,15.359319


In the analysis of loan approval rate, it shows that the platform approves 84.6% of applications, indicating its overall approach.

2. Factors Drive Approval

In [183]:
# Credit Score Affect Approval
df.groupby(df["decision"])["credit_score"].mean()

,credit_score
decision,
Approved,700.924414
Rejected,689.338977


Credit score alone does not significantly differentiate approved and rejected (or risky) borrowers, indicating that lending decisions and loan outcomes depend on multiple factors beyond credit score.

In [184]:
#Let's check the distribution
df.groupby("decision")["credit_score"].describe()

,count,mean,std,min,25%,50%,75%,max
decision,,,,,,,,
Approved,557755.0,700.924414,33.367569,662.0,677.0,692.0,717.0,847.5
Rejected,101212.0,689.338977,25.891270,662.0,672.0,682.0,702.0,847.5


While credit score is an important factor, it alone does not guarantee loan success. Some high-credit borrowers still fall into risky segments, indicating the need for multi-factor risk assessment.

In [185]:
# Loan Approval Rate based on Credit Score, DTI, Annual Income
df.groupby("decision")[["credit_score","dti","annual_inc"]].mean()

,credit_score,dti,annual_inc
decision,,,
Approved,700.924414,19.021721,78783.497191
Rejected,689.338977,20.878088,71139.110635


Approval decisions are influenced by a combination of credit score, income, and debt levels rather than a single factor.